omegaconf/listconfig.py

```diff
@@ -252,12 +255,27 @@ def __setitem__(self, index: Union[int, slice], value: Any) -> None:
             self._format_and_raise(key=index, value=value, cause=e)
 
     def append(self, item: Any) -> None:
+        from omegaconf._utils import is_legal_assignment
+
         try:
             from omegaconf.omegaconf import OmegaConf, _maybe_wrap
 
             index = len(self)
             self._validate_set(key=index, value=item)
 
+            element_type = self.__dict__["_metadata"].element_type
+            if is_container_annotation(element_type) and isinstance(
+                item, BaseContainer
+            ):
+                item_ref_type = item._metadata.ref_type
+                if is_container_annotation(item_ref_type) and not is_legal_assignment(
+                    element_type, item_ref_type
+                ):
+                    _raise_invalid_assignment(element_type, item_ref_type, item)
+                else:
+                    # regular dict without annotation
+                    item = item._value()
+
```

Comment:
I added the function `is_legal_assignment` in utils.py to not re-wrap as you mentioned. Even though we check with `is_legal_assignment`, it's necessary a default case where we re-wrap in case it's a regular dict. For example:

The following snippet where `d`is a DictConfig without a Dict annotation, if there wasn't a default option where it re-wraps the DictConfig, it wouldn't be possible to find if it's a legal assignment.

```python3
@dataclass
class ContainerInList:
    list_with_list: List[List[int]] = field(default_factory=lambda: [[0, 1], [2, 3]])
    list_with_dict: List[Dict[str, int]] = field(
        default_factory=lambda: [{"foo": 1, "foo2": 2}]
    )
d = {"list_with_list": [[4, 5]], "list_with_dict": [{"foo": User()}]}
cfg = OmegaConf.structured(ContainerInList)
OmegaConf.merge(cfg, d) # no error
```

Re-wrapping wouldn't be necessary if somehow we built `ref_type`, `element_type`, `key_type` on runtime for DictConfigs like `d`. I don't know if this is impossible to do.

Reply:
Moreover, I added a test case example in 0b9cbfe which is the only one that covers this example. I'll wait for your review and then, complete with more test cases like the one in 0b9cbfe.